src/compiler/transformers/classFields.ts

```diff
@@ -176,6 +184,77 @@ namespace ts {
                     return visitForStatement(node as ForStatement);
                 case SyntaxKind.TaggedTemplateExpression:
                     return visitTaggedTemplateExpression(node as TaggedTemplateExpression);
+                case SyntaxKind.ThisKeyword:
+                    return visitThisExpression(node as ThisExpression);
+                case SyntaxKind.SuperKeyword:
+                    return visitSuperExpression(node as SuperExpression);
+                case SyntaxKind.FunctionDeclaration:
+                case SyntaxKind.FunctionExpression:
+                case SyntaxKind.Constructor:
+                    return visitNonStaticBoundaryFunction(node as FunctionDeclaration | FunctionExpression | ConstructorDeclaration);
+                case SyntaxKind.GetAccessor:
+                case SyntaxKind.SetAccessor:
+                case SyntaxKind.MethodDeclaration:
+                    return visitNameInStaticBoundaryFunction(node as AccessorDeclaration | MethodDeclaration);
+            }
+            return visitEachChild(node, visitor, context);
+        }
+
+        function visitNameInStaticBoundaryFunction(node: AccessorDeclaration | MethodDeclaration) {
+            function nameInContextVisitor(n: Node): Node {
+                if (node.name === n) {
+                    return visitNode(n, visitor);
+                }
+
+                const savedStaticFieldInitializerContext = staticFieldInitializerContext;
+                staticFieldInitializerContext = undefined;
+                const result = visitNode(n, visitor);
+                staticFieldInitializerContext = savedStaticFieldInitializerContext;
+                return result;
+            }
+            return visitEachChild(node, nameInContextVisitor, context);
+        }
+
+        function visitNonStaticBoundaryFunction(node: FunctionDeclaration | FunctionExpression | ConstructorDeclaration) {
+            const savedStaticFieldInitializerContext = staticFieldInitializerContext;
+            staticFieldInitializerContext = undefined;
+            const result = visitEachChild(node, visitor, context);
+            staticFieldInitializerContext = savedStaticFieldInitializerContext;
+            return result;
+        }
+
+        function visitThisExpression(node: ThisExpression): Expression {

```

Comment:
Is this something we could do in substitution, instead of walking the tree to every `this` or `super` reference during transformation?

You could use `onEmitNode` and enable emit notification to keep track of the outer tree state as the printer walks the tree.

Reply:
Sorry, I'm not familiar with `PipelinePhase` (as well as `Substitution`). Are there any docs or introductions about that?

Reply:
And we a context called staticFieldInitializerContext. And we have visited property declaration.

Do we still need them if we are using `Substitution` ?

Reply:
A tree walk is an expensive operation, especially when we must perform multiple successive tree walks for each transformation (i.e., `ts` -> `esnext` -> `es2020` -> `es2019` -> `es2018` -> etc.). In order to reduce this cost, we build up `TransformFlags` as we construct an AST that provide information about a node and its subtree.

However, some transformations generally would require descending all the way into every `Identifier` or `ThisExpression` in an AST, which can be costly. For example, converting `this` to `_this` in transformed arrow functions, or `foo` into `some_module_1.foo` for an import, etc. To avoid that expensive walk, we have a separate transformation phase called "Just-in-Time" (JIT) substitution. JIT substitution is invoked by the emitter when the entire AST is walked as we print, since we're guaranteed to visit every node as we descend through the tree.

JIT substitution is broken down into two components:
- Emit Notification
- Substitution

## Emit Notification

With any regular transformation, we generally use the call stack to save and restore context information relevant to children, such as with `staticFieldInitializerContext`:

```ts
function visitSomeNode(node: SomeNode) {
    // save old state as we descend
    const savedState = someState;
    // set new state for subtree
    someState = ...;
        
    const newNode = factory.updateSomeNode(node, ...);
        
    // restore old state
    someState = savedState;

    return newNode;
}
```

When we perform JIT substitution, we often need the same capability. To accommodate this need, the emitter will trigger "emit notifications" that can be set up via the `TransformationContext`:

```ts
export function transformSomething(context: TransformationContext) {
    const previousOnEmitNode = context.onEmitNode; // capture existing notification hook
    context.onEmitNode = onEmitNode; // overwrite the notification hook

    ...

    function onEmitNode(hint: EmitHint, node: Node, emitCallback: (hint: EmitHint, node: Node) => void) {
        // save old state as we descend
        const savedState = someState;
        // set new state for subtree
        someState = ...;

        // re-enter the emitter
        emitCallback(hint, node);

        // restore old state
        someState = savedState;
    }
}
```

Emit notification doesn't necessarily trigger for all `Node`s, however, since that too would be expensive. Instead, you must inform the emitter than emit notification is necessary either for any node with a specific `SyntaxKind`, or by setting an `EmitFlag` on a specific node:

```ts
export function transformSomething(context: TransformationContext) {
    const previousOnEmitNode = context.onEmitNode; // capture existing notification hook
    context.onEmitNode = onEmitNode; // overwrite the notification hook
    context.enableEmitNotification(SyntaxKind.SomeNode);
    ...
}

// or
setEmitFlags(node, EmitFlags.AdviseOnEmitNode);
```

If you need to share state between a regular transformation and a later substitution, we generally do this by creating something like a `Map` in the transformer, and using that to store information for later reuse by using the original node id:

```ts
export function transformSomething(context: TransformationContext) {
    const previousOnEmitNode = context.onEmitNode; // capture existing notification hook
    context.onEmitNode = onEmitNode; // overwrite the notification hook
    context.enableEmitNotification(SyntaxKind.SomeNode);

    let someState: SomeState | undefined;
    const someStateForNode = new Map<number, SomeState>();

    ...

    function visitSomeNode(node: SomeNode) {
        const savedState = someState;
        someState = ...;

        const newNode = factory.updateSomeNode(node, ...);

        // persist state for later substitution
        someStateForNode.set(getOriginalNodeId(node), someState);

        someState = savedState;
        return newNode;
    }

    ...

    function onEmitNode(hint: EmitHint, node: Node, emitCallback: (hint: EmitHint, node: Node) => void) {
        if (node.kind === SyntaxKind.SomeNode) {
            const savedState = someState;
            someState = someStateForNode.get(getOriginalNodeId(node));

            // re-enter the emitter
            emitCallback(hint, node);

            // restore old state
            someState = savedState;
        }
        else {
            emitCallback(hint, node);
        }
    }
}
```

This essentially allows us to restore the necessary surrounding context for a substitution, though its important to keep in mind that the `Node` we encounter during `onEmitNode` may not be the same `Node` we saw during transformation due to subsequent transformations (i.e., an `ArrowFunction` may have been transformed into a `FunctionExpression`). This is why we use `getOriginalNodeId`, since this gets the `id` of the "original" `Node`, a pointer that is set when performing an `factory.updateX` function or manually using `setOriginalNode`.

## Substitution

Once you have established the correct context via emit notification, you can set inject a substitution hook in the same way as an emit notification hook:

```ts
export function transformSomething(context: TransformationContext) {
    const previousOnSubstituteNode = context.onSubstituteNode; // capture existing substitution hook
    context.onSubstituteNode = onSubstituteNode; // overwrite the substitution hook
    context.enableSubstitution(SyntaxKind.ThisKeyword);

    ...

    function onSubstituteNode(hint: EmitHint, node: Node) {
        // first, evaluate existing substitutions
        node = previousOnSubstituteNode(hint, node);

        // now perform our own
        if (node.kind === SyntaxKind.ThisKeyword) {
            return substituteThisExpression(node as ThisExpression);
        }

        return node;
    }

    function substituteThisExpression(node: ThisExpression) {
        // use the state we established via emit notification
        ...
    }
}
```

Reply:
Thanks for the explanation.
In my rough understanding, this is looks like merging multiple AST traversals into once?

Reply:
Another problem is what if the node has already been transformed.
For example:

```ts
class C {
    static a = 1
}

```

to

```ts
var C = /** @class */ (function () {
    function C() {
    }
    C.a = 1;
    return C;
}());

```

Should I enable `Emit Notification` and `Substitute` for the `FunctionDeclaration` or `ClassDeclaration`?

Reply:
Friendly ping. needs some suggestions here😂.

Reply:
It may not be necessary. You could wire up an `original` pointer to the original PropertyDeclaration on the `C.a = 1` expression using `setOriginalNode`, as well as call `addEmitFlags(node, EmitFlags.AdviseOnEmitNode)`. That will cause emit notification to be triggered for that specific node. Then in your `onEmitNode` hook you can look test for the property declaration:

```ts
function onEmitNode(hint, node, emitCallback) {
  if (isAssignmentExpression(node, true)) {
    const original = getOriginalNode(node);
    if (isPropertyDeclaration(original)) {
      // we have the original property declaration.
      ...
    }
  }
}
```

Note that an "original" node is not guaranteed to be a "parse tree node", and only "parse tree nodes" are guaranteed to have a `parent`. You _could_ use `getParseTreeNode` instead of `getOriginalNode`, but that means that this would break if we later add a transformation that introduces a synthetic property declaration for any reason. To avoid that, you might want to associate the `staticFieldInitializerContext` in `staticFieldInitializerContextMap` with the node id of each `PropertyDeclaration`, rather than just the class itself.

Reply:
It's a little bit hard if we are enabled useDefineField. Because It's generated a ObjectDefine call.